### PR TITLE
attestation: improve error message when `gh` is too old

### PR DESCRIPTION
**Pull request by nandahkrishna**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Currently, when `gh` is too old to verify attestations, the error message is uninterpretable:
```
==> Installing frizbee dependency: go
==> Downloading https://ghcr.io/v2/homebrew/core/go/manifests/1.22.5
==> Verifying attestation for go
Error: The bottle for go has an invalid build provenance attestation.

This may indicate that the bottle was not produced by the expected
tap, or was maliciously inserted into the expected tap's bottle
storage.

Additional context:

attestation verification failed: Failure while executing; `/usr/bin/env GH_TOKEN=****** /opt/homebrew/bin/gh attestation verify /Users/REDACTED/Library/Caches/Homebrew/downloads/3f2258381d80802127c7c7d253acee16b65864703253acd14e6e63e6194b0b62--go--1.22.5.arm64_sonoma.bottle.tar.gz --repo trailofbits/homebrew-brew-verify --format json` exited with 1. Here's the output:
Failed to verify the artifact: failed to fetch attestations for subject: sha256:555bf0e72c91ff434d4ab9e69c001cd998f6ce23fdddcac7013f94343d0defda
```

This PR aims to improve the messaging, like below:
```
➜  homebrew git:(attestation-gh-error) ✗ HOMEBREW_NO_INSTALL_FROM_API=1 brew install -sdv frizbee
[TRUNCATED]
==> Installing dependencies for frizbee: go
==> Installing frizbee dependency: go
==> Downloading https://ghcr.io/v2/homebrew/core/go/manifests/1.22.5
==> Verifying attestation for go
Error: `gh` is too old, you must run `brew upgrade gh` to continue.
```

### CODE REVIEW

File: Library/Homebrew/attestation.rb [hunk 1]
```diff
@@ -104,6 +107,11 @@ def self.check_attestation(bottle, signing_repo, signing_workflow = nil, subject
         # Even if we have credentials, they may be invalid or malformed.
         raise GhAuthNeeded, "invalid credentials" if e.status.exitstatus == 4
 
+        if Formula["gh"].any_installed_version.version < GH_ATTESTATION_MIN_VERSION
```
Comment (ZhongRuoyu) [inline Library/Homebrew/attestation.rb:110]: We should probably check `gh --version` instead. `gh_executable` may or may not be installed with `brew`.

Comment (nandahkrishna) [reply Library/Homebrew/attestation.rb:110]: That would be better, but we currently don't verify attestations unless `gh` is installed from Homebrew. See here:
https://github.com/Homebrew/brew/blob/2d26c89e81e838f34927c89d9d368765e9dc56df/Library/Homebrew/attestation.rb#L52

I'll make the change though!

Comment (woodruffw) [reply Library/Homebrew/attestation.rb:110]: We do _if_ the user explicitly enables attestations via `HOMEBREW_VERIFY_ATTESTATIONS` 🙂 -- so yeah, we do need this to use `--version` instead.

Comment (nandahkrishna) [reply Library/Homebrew/attestation.rb:110]: Ah, my bad – updating this. Thanks!

File: Library/Homebrew/attestation.rb [hunk 2]
```diff
@@ -104,6 +107,13 @@ def self.check_attestation(bottle, signing_repo, signing_workflow = nil, subject
         # Even if we have credentials, they may be invalid or malformed.
         raise GhAuthNeeded, "invalid credentials" if e.status.exitstatus == 4
 
+        gh_version = Version.new(system_command!(gh_executable, args: ["--version"], print_stderr: false)
+                                 .stdout.match(/\d+(?:\.\d+)+/i).to_s)
+        if gh_version < GH_ATTESTATION_MIN_VERSION
+          raise e,
+                "`gh` is too old, you must run `brew upgrade gh` to continue."
```
Comment (MikeMcQuaid) [inline Library/Homebrew/attestation.rb:114]:
```suggestion
                "`#{gh_executable}` is too old, you must run upgrade it to continue."
```

maybe we should automatically upgrade it if it's an installed formula?

**Review by MikeMcQuaid — COMMENTED** [review]
One note then good to 🚢!

**Review by MikeMcQuaid — APPROVED** [review]

File: Library/Homebrew/attestation.rb [hunk 3]
```diff
@@ -63,7 +66,7 @@ def self.gh_executable
       #       to prevent a cycle during bootstrapping. This can eventually be resolved
       #       by vendoring a pure-Ruby Sigstore verifier client.
       @gh_executable ||= T.let(with_env("HOMEBREW_VERIFY_ATTESTATIONS" => nil) do
```
Comment (woodruffw) [inline Library/Homebrew/attestation.rb:68]: While we're at it: let's flip this to:

```suggestion
      @gh_executable ||= T.let(with_env("HOMEBREW_NO_VERIFY_ATTESTATIONS" => "1") do
```

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by nandahkrishna** [comment]
I've ignored bootstrapping the latest `gh` or trying to upgrade it because it's a little too complicated for what it's worth. Now this just raises an error if `gh` is too old, and asks the user to upgrade it.

File: Library/Homebrew/attestation.rb [hunk 4]
```diff
@@ -59,10 +62,10 @@ def self.enabled?
     # @api private
     sig { returns(Pathname) }
     def self.gh_executable
-      # NOTE: We disable HOMEBREW_VERIFY_ATTESTATIONS when installing `gh` itself,
+      # NOTE: We disable HOMEBREW_NO_VERIFY_ATTESTATIONS when installing `gh` itself,
```
Comment (woodruffw) [inline Library/Homebrew/attestation.rb:65]:
```suggestion
      # NOTE: We set HOMEBREW_NO_VERIFY_ATTESTATIONS when installing `gh` itself,
```

File: Library/Homebrew/attestation.rb [hunk 5]
```diff
@@ -104,6 +107,13 @@ def self.check_attestation(bottle, signing_repo, signing_workflow = nil, subject
         # Even if we have credentials, they may be invalid or malformed.
         raise GhAuthNeeded, "invalid credentials" if e.status.exitstatus == 4
 
+        gh_version = Version.new(system_command!(gh_executable, args: ["--version"], print_stderr: false)
+                                 .stdout.match(/\d+(?:\.\d+)+/i).to_s)
+        if gh_version < GH_ATTESTATION_MIN_VERSION
+          raise e,
+                "#{gh_executable} is too old, you must upgrade it to continue."
```
Comment (woodruffw) [inline Library/Homebrew/attestation.rb:114]:
```suggestion
                "#{gh_executable} is too old, you must upgrade it to continue"
```

File: Library/Homebrew/test/attestation_spec.rb [hunk 6]
```diff
@@ -74,10 +77,42 @@
     end
   end
 
+  describe "::check_attestation (with old gh)" do
```
Comment (woodruffw) [inline Library/Homebrew/test/attestation_spec.rb:80]:
```suggestion
  describe "::check_attestation fails with old gh" do
```

**Review by woodruffw — APPROVED** [review]

**Pull request merged**